### ISSUE TITLE
Extract-method C7 crash

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#29536]
i was trying to extract the following code:

```c#
                    if (method != null &&
                        method.DeclaredAccessibility == Accessibility.Public &&
                        method.IsStatic &&
                        _regexType.Equals(method.ContainingType))
                    {
                        return AnalyzeStringLiteral(
                            stringLiteral, argumentNode, cancellationToken, out options);
                    }

                    return true;
```

Callstack:

```
System.InvalidOperationException : Matrix does not contain Key 'C7'.
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition,String message)
   at Microsoft.CodeAnalysis.ExtractMethod.ExtractMethodMatrix.GetVariableStyle(Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GetVariableStyle(Dictionary`2 symbolMap,ISymbol symbol,SemanticModel model,ITypeSymbol type,Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GenerateVariableInfoMap(SemanticModel model,DataFlowAnalysis dataFlowAnalysisData,Dictionary`2 symbolMap)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.AnalyzeAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
```

**Comment by CyrusNajmabadi** [comment]
Note: if i remove hte 'out options' part, then things work fine.  so it has something to do with an out-var.

**Comment by CyrusNajmabadi** [comment]
tagging @jinujoseph @heejaechang

**Comment by alrz** [comment]
I thought we were going to just not trying to offer the option instead of crashing here. What is the rationale behind this flat out halt when there's a possible unhandled case?

**Comment by CyrusNajmabadi** [comment]
agreed.  @jinujoseph extract method has a pretty impactful problem that we get tons of hits over.  Effectively, it has a whitelist of things it supports but then *throws* when it encounters something it doesn't know about.

This is pretty strange, and is unlike every other analyzer/fix we have.  If we want to collect information about people hitting things we do not support, we can do that through other supported means (like telemetry).  Having the feature crash seems strange.

**Comment by jinujoseph** [comment]
Thought this was fixed [here](https://github.com/dotnet/roslyn/issues/23352)
looks like have more to do here

**Comment by CyrusNajmabadi** [comment]
@jinujoseph Unfortunately, that fix only expanded extract-method's whitelist.  It didn't change the general behavior of crashing when it encounters something it doesn't know about.  That's the part of extract-method we should change.

Thanks!

**Comment by heejaechang** [comment]
is it actual crash or info bar? it already send telemetry rather than crashing.

**Comment by verelpode** [comment]
This error happened to me again today with VS 15.8.3.  I didn't explicitly click any command to extract method, rather the error popped up by itself while I was typing some code.  
I presume that I won't see this error again because the message says it has been disabled and I didn't click the Enable button (because I've seen it enough times and would prefer not to see it again).

```
ExtractMethodCodeRefactoringProvider encountered an error and has been disabled.

System.InvalidOperationException : Matrix does not contain Key 'C7'.
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition,String message)
   at Microsoft.CodeAnalysis.ExtractMethod.ExtractMethodMatrix.GetVariableStyle(Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GetVariableStyle(Dictionary`2 symbolMap,ISymbol symbol,SemanticModel model,ITypeSymbol type,Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GenerateVariableInfoMap(SemanticModel model,DataFlowAnalysis dataFlowAnalysisData,Dictionary`2 symbolMap)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.AnalyzeAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)

```

**Comment by jinujoseph** [comment]
also reported [here](https://developercommunity.visualstudio.com/content/problem/329061/extractmethodcoderefactoringprovider-error-1.html) in 15.8.2

**Comment by verelpode** [comment]
I upgraded to VS 15.8.5 but this "C7" error happened to me again today with VS 15.8.5.  I don't know what triggers it, but it happens while I'm writing C# code.

```
System.InvalidOperationException : Matrix does not contain Key 'C7'.
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition,String message)
   at Microsoft.CodeAnalysis.ExtractMethod.ExtractMethodMatrix.GetVariableStyle(Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GetVariableStyle(Dictionary`2 symbolMap,ISymbol symbol,SemanticModel model,ITypeSymbol type,Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GenerateVariableInfoMap(SemanticModel model,DataFlowAnalysis dataFlowAnalysisData,Dictionary`2 symbolMap)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.AnalyzeAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

**Comment by verelpode** [comment]
> Effectively, it has a whitelist of things it supports but then throws when it encounters something it doesn't know about.  This is pretty strange

I agree, that is unfeasible behavior, because at any time when someone is in the middle of writing some C# code/method/class but hasn't finished writing it, then obviously the code is invalid simply because it's unfinished, and the number of combinations of unfinished/invalid code are **infinite**.  You cannot make a whitelist nor blacklist cover an **endless** number of possibilities.

**Comment by heejaechang** [comment]
@verelpode that's not what it whitelist. it has expected data flow/control flow result and with new language features that got changed but the expectation assert hasn't been changed. that's why you see info bar.

we need to take a look but got delayed since it is not actual crash but info bar. but will take a look. thank you for the comment!

**Comment by verelpode** [comment]
@heejaechang Aha, OK, thanks.  By the way, the error appears even though I did not click the "Extract Method" command.  I suggest that if possible the "Extract Method" code should only run when the user explicitly clicks the "Extract Method" command.  If that is impossible, then at least minimize the amount of code that runs _without_ the user clicking the "Extract Method" command.

**Comment by heejaechang** [comment]
root cause - https://github.com/dotnet/roslyn/issues/30548

**Comment by CyrusNajmabadi** [comment]
Fixed and mitigated with @heejaechang's and my changes.

**Comment by verelpode** [comment]
This C7 problem was supposed to be fixed, but surprisingly it happened to me again today with VS 15.9.3.  Following is the code that caused it (with some names changed).

```
		private MyEnum1 _GetValueValidity(StringSegment inValueStr, out StringSegment outCorrectedString, bool inAllowTrim)
		{
			outCorrectedString = inValueStr;
			if (!fIsSurroundingWhiteSpaceAllowed)
			{
				StringSegment trimmedStrSeg = inValueStr.Trim();
				if (trimmedStrSeg.Length < inValueStr.Length)   // xxxxx
				{
					if (!inAllowTrim) return MyEnum1.XXX1;
					inValueStr = trimmedStrSeg;
					outCorrectedString = trimmedStrSeg;
				}
			}

			if (!fAreLineBreakCharsAllowed && MyHelper1.FindVertWhiteSpace(inValueStr) >= 0)
				return MyEnum1.XXX2;
			
			if (fAllowedCharMatcher != null && fAllowedCharMatcher.FindNonChar(inValueStr) >= 0)
				return MyEnum1.XXX1;

			int valueStrLength = inValueStr.Length;
			if (valueStrLength < fMinCharCount)
				return (valueStrLength == 0) ? MyEnum1.XXX3 : MyEnum1.XXX4;
			int mxcc = fMaxCharCount;
			if (mxcc >= 0 && valueStrLength > mxcc)
				return MyEnum1.XXX5;

			return MyEnum1.XXX6;
		}

		public override MyEnum1 GetValueValidity()
		{
			return this._GetValueValidity(fEditTextBox.Text, inAllowTrim: true);
		}
```

```
ExtractMethodCodeRefactoringProvider encountered an error and has been disabled.
System.InvalidOperationException : Matrix does not contain Key 'C7'.
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition,String message)
   at Microsoft.CodeAnalysis.ExtractMethod.ExtractMethodMatrix.GetVariableStyle(Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GetVariableStyle(Dictionary`2 symbolMap,ISymbol symbol,SemanticModel model,ITypeSymbol type,Boolean captured,Boolean dataFlowIn,Boolean dataFlowOut,Boolean alwaysAssigned,Boolean variableDeclared,Boolean readInside,Boolean writtenInside,Boolean readOutside,Boolean writtenOutside,Boolean unsafeAddressTaken)
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GenerateVariableInfoMap(SemanticModel model,DataFlowAnalysis dataFlowAnalysisData,Dictionary`2 symbolMap)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.AnalyzeAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

**Comment by CyrusNajmabadi** [comment]
> This C7 problem was supposed to be fixed, but surprisingly it happened to me again today with VS 15.9.3. 

I don't believe the fixes went into 15.9.  You'll need 16.0+
